### PR TITLE
Categorical prediction implemented (Debug needed)

### PR DESCRIPTION
**Pull request by bulaienTang**
Changed three files: 

1. [climate_learn/models/components/resnet.py](https://github.com/tung-nd/climate_learn/compare/categorical-prediction?expand=1#diff-b552c7d1c5804daee999d71a3418a7a04343b86cc9dc27d678fdc51e5626be1c)
  - made changes to pred function when prob_type = 'categorical', basically following the implementation on https://github.com/sagar-garg/WeatherBench/blob/f41f497ac45377d363dc30bfa77daf50d7b28afd/src/networks.py#L208

2. [climate_learn/models/modules/forecast.py](https://github.com/tung-nd/climate_learn/compare/categorical-prediction?expand=1#diff-451f4ad38bb86ea815a28a320e7b9370f22479ff2a636e683ae2d018886fdc0f)
 -  made categorical specific changes on training/validation/test steps, and transformed y to be one-hot encoded, based from https://github.com/sagar-garg/WeatherBench/blob/f41f497ac45377d363dc30bfa77daf50d7b28afd/src/data_generator.py#L335

3. [climate_learn/models/modules/utils/metrics.py](https://github.com/tung-nd/climate_learn/compare/categorical-prediction?expand=1#diff-caeaad061a6b7531de70885c49b77a2f66d8ff89fde84640278004ce75f46068)
 - created categorical loss, but might be buggy. Consider the original implementation: https://github.com/sagar-garg/WeatherBench/blob/f41f497ac45377d363dc30bfa77daf50d7b28afd/src/networks.py#L423

### CODE REVIEW

File: climate_learn/models/components/resnet.py [hunk 1]
```diff
@@ -82,15 +82,32 @@ def __init__(
         if prob_type == 'parametric':
             self.final_std = PeriodicConv2D(hidden_channels, out_channels, kernel_size=7, padding=3)
 
-    def predict(self, x):
+        # final layer for categorical, change to be number of bins
+        if self.prob_type == 'categorical':            
+            self.final = PeriodicConv2D(self.hidden_channels, 50, kernel_size=7, padding=3)
```
Comment (jasonjewik) [inline climate_learn/models/components/resnet.py:87]: The second argument here should be parameterized that can be passed in at initialization (e.g., `self.num_output_bins`).

File: climate_learn/models/components/resnet.py [hunk 2]
```diff
@@ -206,15 +225,19 @@ def test_rollout(self, x: torch.Tensor, y: torch.Tensor, clim, variables, out_va
...
-        preds = []
-        for _ in range(steps):
-            x = self.predict(x)
-            if self.prob_type == 'parametric':
-                x = mean_transform(x.loc)
-            preds.append(x)
-        preds = torch.stack(preds, dim=1)
-        if len(y.shape) == 4:
-            y = y.unsqueeze(1)
+        # need debug here!! what to do with categorical and other prob_type?
```
Comment (jasonjewik) [inline climate_learn/models/components/resnet.py:228]: Can you elaborate on this? What is the issue?

File: climate_learn/models/modules/forecast.py [hunk 3]
```diff
@@ -39,6 +42,16 @@ def __init__(
             self.train_loss = lat_weighted_mse
             self.val_loss = [crps_gaussian_val]
             raise NotImplementedError("Only parametric and deterministic prediction is supported")
+        elif net.prob_type == 'categorical':
+            # loss functions need to be determined later (?)
```
Comment (jasonjewik) [inline climate_learn/models/modules/forecast.py:46]: These loss functions lgtm

**Review by jasonjewik — CHANGES_REQUESTED** [review]

**Comment by jasonjewik** [comment]
Can you also apply these changes to the other model components like CNN block, Unet, ViT?

File: climate_learn/models/components/resnet.py [hunk 4]
```diff
@@ -42,7 +42,7 @@ def __init__(
         else:
             raise NotImplementedError(f"Activation {activation} not implemented")
         
-        assert not prob_type or prob_type in ['parametric', 'mcdropout', 'deter']
+        assert not prob_type or prob_type in ['parametric', 'mcdropout', 'deter', 'categorical']
```
[the remark this answers is not in the document]
Comment (bulaienTang) [reply climate_learn/models/components/resnet.py:45]: Should I remove all of its occurrences then?

File: climate_learn/models/components/resnet.py (hunk 1 above)
Comment (bulaienTang) [reply climate_learn/models/components/resnet.py:87]: resolved.

File: climate_learn/models/components/resnet.py (hunk 2 above)
Comment (bulaienTang) [reply climate_learn/models/components/resnet.py:228]: So the original code for this part was everything inside the else statement (from line 232 to 240) and when I ran the categorical, errors popped up, and I wasn't able to fix them at that time, so I made an if statement and simply ran`preds = self.predict(x, len(out_variables))` for categorical. 

And I'm bit confused why did the val_rollout has different implementations for probability type and regular deterministic methods. while for test rollout, the original code doesn't have different implementations?

File: climate_learn/models/components/resnet.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jasonjewik) [reply climate_learn/models/components/resnet.py:45]: I would recommend it.

File: climate_learn/models/components/resnet.py (hunk 2 above)
Comment (jasonjewik) [reply climate_learn/models/components/resnet.py:228]: It doesn't look like we have a good reason for why val_rollout is different from test_rollout. I dug through our commit history and it looks like they just ended up that way because of the uncertainty repo merge: https://github.com/aditya-grover/climate-learn/blob/300e602b06072c3d9e8e19841c9526998debb3c7/climate_learn/models/components/resnet.py. 

So IMO we have the flexibility to change the rollouts to match each other. I think it makes sense to have them be the same. Can you see what happens if you make the rollouts match? Does our demo notebook still run?

Comment (bulaienTang) [reply climate_learn/models/components/resnet.py:228]: Looks like it's working

Comment (jasonjewik) [reply climate_learn/models/components/resnet.py:228]: Great, thanks! Once you resolve the merge conflicts, I will approve this PR.

**Review by jasonjewik — COMMENTED** [review]
One more thing: can you delete `TestCategorical.ipynb`? We should not be including test notebooks in the repo.

**Comment by bulaienTang** [comment]
I think it's reading for merge now.

**Review by jasonjewik — APPROVED** [review]
lgtm
